### PR TITLE
Cleanup/Optimize gas for swaps

### PR DESCRIPTION
**Pull request by stackman27**
Closes: [#5842 ](https://github.com/osmosis-labs/osmosis/issues/5842)

## What is the purpose of the change

- [x] Theres one trivial extra GetPool call we can eliminate from all swaps by making compute swap methods take in a pool instead of poolID

- [x] I think its maybe two GetPools eliminatable, including the one in updatePoolForSwap? This one not as clear

- [x] Similarly we can use the multi-send api's from bank to optimize the final fund transfers

## Testing and Verifying
tested on localosmosis by performing swap and comparing outputs 

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/swaps.go [hunk 1]
```diff
@@ -660,39 +671,67 @@ func (k Keeper) updatePoolForSwap(
...
+	// 	},
+	// }
+
+	// err := k.bankKeeper.InputOutputCoins(ctx, inputs, outputs)
+	// if err != nil {
+	// 	return types.InsufficientPoolBalanceError{Err: err}
+	// }
+
+	// Send the input token from the user to the pool's primary address
+	err := k.bankKeeper.SendCoins(ctx, swapDetails.Sender, pool.GetAddress(), sdk.Coins{
```
Comment (stackman27) [inline x/concentrated-liquidity/swaps.go:723]: note: i tried using [multisend api](https://github.com/osmosis-labs/osmosis/pull/5850/files#diff-014d8f3d7938e869988af2d7f7ada29412dd1c7490e7cbcd83fa4f3903ef7e09R691-R720) to replace multiple sendCoins call, but the keeper function doesnot take multiple input address 
ref: https://github.com/osmosis-labs/cosmos-sdk/blob/osmosis-main/x/bank/types/msgs.go#L174-L175

Comment (ValarDragon) [reply x/concentrated-liquidity/swaps.go:723]: Right, we will want to still do multiple multi-sends, just only one send per sender!

File: x/concentrated-liquidity/swaps.go [hunk 2]
```diff
@@ -660,39 +673,52 @@ func (k Keeper) updatePoolForSwap(
...
-		swapDetails.TokenOut,
-	})
+	err = k.bankKeeper.InputOutputCoins(ctx, []banktypes.Input{{
+		Address: pool.GetAddress().String(),
+		Coins:   sdk.NewCoins(swapDetails.TokenOut),
+	}}, []banktypes.Output{
+		{
+			Address: swapDetails.Sender.String(),
+			Coins:   sdk.NewCoins(swapDetails.TokenOut),
+		}})
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:720]: What is the benefit of using multi-send API when we replace each regular API with a multi-send one?

I would assume that we would want to batch all inputs and all outputs and run a single multi-send API at the end.

File: x/concentrated-liquidity/swaps_test.go [hunk 3]
```diff
@@ -3404,10 +3437,87 @@ func (s *KeeperTestSuite) TestInfiniteSwapLoop_OutGivenIn() {
...
+
+	s.Require().True(s.Ctx.GasMeter().GasConsumed() < maxGasToConsume)
+
+	pool, err = s.clk.GetPoolById(s.Ctx, pool.GetId())
+	s.Require().NoError(err)
+
+	// Make a large swap in InGivenOut (negative tick direction) from currentTick = 1735 to 736
+	// Amount of tickCrossed = 998
+	// Make sure the gas is < 50M
+	_, _, _, err = s.clk.SwapInAmtGivenOut(s.Ctx, swapAddress, pool, sdk.NewCoin(USDC, sdk.NewInt(50_000_000_000)), ETH, pool.GetSpreadFactor(s.Ctx), sdk.ZeroDec())
```
Comment (stackman27) [inline x/concentrated-liquidity/swaps_test.go:3519]: ref: 
swapGasAmount(sendCoins): 22_367_874
swapGasAmount(InputOutputCoins): 22_345_380

File: x/concentrated-liquidity/swaps.go (hunk 2 above)
Comment (stackman27) [reply x/concentrated-liquidity/swaps.go]: yea i think so, using InputOutputCoins made gas slightly lower. ref: https://github.com/osmosis-labs/osmosis/pull/5850/files#r1266377932

**Comment by ValarDragon** [comment]
This is state breaking and can't be backported due to affecting gas.

**Comment by ValarDragon** [comment]
Can you add a changelog for improving gas of CL swaps?

File: x/concentrated-liquidity/swaps.go (hunk 2 above)
Comment (ValarDragon) [reply x/concentrated-liquidity/swaps.go:724]: This change here isn't quite what I had in mind. I was expecting to be doing a MultiSend (not an InputOutputCoins!) and combining the two sends from the sender! My bad, I forgot that MultiSend is the name of the SDK message, which doesn't actually use the sendManyCoins API.

https://github.com/osmosis-labs/cosmos-sdk/blob/osmosis-main/x/bank/keeper/send.go#L21C7-L21C7

Comment (ValarDragon) [reply x/concentrated-liquidity/swaps.go:724]: Though this suggests an interesting idea, thats actually better than mine.

Could we do all the sends here as a single InputOutputCoins call?

**Comment by stackman27** [comment]
> Can you add a changelog for improving gas of CL swaps?

added changelog

File: x/concentrated-liquidity/fuzz_test.go [hunk 4]
```diff
@@ -278,6 +278,9 @@ func (s *KeeperTestSuite) swap(pool types.ConcentratedPoolExtension, swapInFunde
 	// // Execute swap
 	fmt.Printf("swap in: %s\n", swapInFunded)
 	cacheCtx, writeOutGivenIn := s.Ctx.CacheContext()
+	pool, err := s.clk.GetPoolById(s.Ctx, pool.GetId())
+	s.Require().NoError(err)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/fuzz_test.go:282]: Why is this needed in place of the pool object that is provided in the args? No state change occurs prior to this call on the pool obj as far as I see.

File: x/concentrated-liquidity/fuzz_test.go [hunk 5]
```diff
@@ -297,6 +300,9 @@ func (s *KeeperTestSuite) swap(pool types.ConcentratedPoolExtension, swapInFunde
 	// We expect the returned amountIn to be roughly equal to the original swapInFunded.
 	cacheCtx, _ = s.Ctx.CacheContext()
 	fmt.Printf("swap out: %s\n", tokenOut)
+	pool, err = s.clk.GetPoolById(s.Ctx, pool.GetId())
+	s.Require().NoError(err)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/fuzz_test.go:304]: Same question as above

File: x/concentrated-liquidity/position_test.go [hunk 6]
```diff
@@ -2426,7 +2426,9 @@ func (s *KeeperTestSuite) TestTickRoundingEdgeCase() {
 	swapAddr := testAccs[2]
 	desiredTokenOut := sdk.NewCoin(USDC, sdk.NewInt(10000))
 	s.FundAcc(swapAddr, sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(1000000000000000000))))
-	_, _, _, err := s.clk.SwapInAmtGivenOut(s.Ctx, swapAddr, pool, desiredTokenOut, ETH, sdk.ZeroDec(), sdk.ZeroDec())
+	pool, err := s.clk.GetPoolById(s.Ctx, pool.GetId())
+	s.Require().NoError(err)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/position_test.go:2430]: Same

File: x/concentrated-liquidity/swaps.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/swaps.go:724]: This still does not make sense to me, this still looks like three separate bank sends to me.

File: x/concentrated-liquidity/swaps_test.go [hunk 7]
```diff
@@ -2091,6 +2091,9 @@ func (s *KeeperTestSuite) TestComputeAndSwapOutAmtGivenIn() {
 				s.assertPoolNotModified(poolBeforeCalc)
 			}
 
+			pool, err = s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, pool.GetId())
+			s.Require().NoError(err)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/swaps_test.go:2095]: Same question with refetching these pool objects again, if we need to refetch multiple times why cant we just operate on the pointer?

File: x/concentrated-liquidity/fuzz_test.go (hunk 4 above)
Comment (stackman27) [reply x/concentrated-liquidity/fuzz_test.go:282]: So pool obj change does when we create first position. It updates the current Sqr Price https://github.com/osmosis-labs/osmosis/blob/main/x/concentrated-liquidity/lp.go#L476

File: x/concentrated-liquidity/swaps_test.go (hunk 7 above)
Comment (stackman27) [reply x/concentrated-liquidity/swaps_test.go:2095]: we refetch because the pool obj change after fist position creation. And all most of these calls are in different test functions, with their own independent logic so i'm not sure how helpful pointer will be

File: x/concentrated-liquidity/fuzz_test.go (hunk 5 above)
Comment (stackman27) [reply x/concentrated-liquidity/fuzz_test.go:304]: answered https://github.com/osmosis-labs/osmosis/pull/5850#discussion_r1273982708

File: x/concentrated-liquidity/position_test.go (hunk 6 above)
Comment (stackman27) [reply x/concentrated-liquidity/position_test.go:2430]: answered https://github.com/osmosis-labs/osmosis/pull/5850#discussion_r1273982708

**Comment by stackman27** [comment]
okay just tried to refactor and there was lots of unintended behaviors in core swap logic around state get/set. So will probably close this PR to not break the current flow
